content/features/featuresDeepDive/smartFilters/howSmartFiltersWork.md

```diff
@@ -0,0 +1,141 @@
+
+## Design Principles
+
+1. Keep the Smart Filter core simple and ensure the runtime is fast
+2. Provide an easy to use API surface
+3. Support only loading the code you need, and support delay loading that code (e.g. deserialization is async so deserialization code doesn't have to be loaded until it is used)
+
+### Smart Filter
+
+The notion of a `SmartFilter` is a graph of blocks (all inheriting from `BaseBlock`) linked to each other through `ConnectionPoint`s.
+
+During the initialization phase, the `SmartFilter` examines the graph and builds a list of commands to execute during each frame and stores them in a `CommandBuffer`:
+
+- This keeps each frame render very fast by minimizing the amount of code running and branches in that code.
+- Basically, every `BaseBlock` during its `generateCommandsAndGatherInitPromises` step will register in the `CommandBuffer` only the work it has to execute according to its options.
+- In the end the render loop only needs to execute each command in the list.
+- This also allows for easy injection of debug tools or logging in between each command without growing the minimum required code to run a filter.
+- Each Block is responsible for "injecting" its command in the Smart Filter's `CommandBuffer`.
+- For example: the `ShaderBlock` highlights how convenient it can be to only switch once at init time between rendering to the main render target (if linked to the output) or to an intermediate texture if in the middle of the chain.
+
+Each block is responsible for exposing their inputs and outputs through `registerInput` and `registerOutput`:
+
+- This creates and stores the related `ConnectionPoint`.
+- Any property values of a block must be updated only before `initialize` is called as the runtime won't reference it at render time.
+- All values that need to be dynamically updated after `initialize` must be defined as connection points or 'StrongRef'.
+- By convention, blocks defined in TypeScript should expose their input connection points as properties for convenience.
+
+The `ConnectionPoints` are:
+
+- Strongly typed, allowing more type safety while creating `SmartFilter` by code, yet keeping enough flexibility to be understood efficiently at runtime.
+- They are also responsible for their compatibility when linked to each other.
+
+The Smart Filter is fully abstracted away from the runtime and does not even require a Babylon Engine to work with. It is only responsible to hold the "graph" of the Smart Filter or the "map" of all its blocks. A Smart Filter only needs a name to be created:
+
+```typescript
+const smartFilter = new SmartFilter("Simplest");
+```
+
+Once a Smart Filter exists, one can add as many blocks as necessary to it:
+
+```typescript
+const blur = new BlurBlock(smartFilter, "blur");
+```
+
+Then, the various blocks can be linked together:
+
+```typescript
+videoInput.output.connectTo(blur.input);

```

Comment:
Is there a missing line for creating the `videoInput`, or is that just supposed to be like a hypothetical block?